### DISCUSSION TITLE
Initializer lists of value types

### DISCUSSION THREAD

**Discussion by Git-i** [hsutter/cppfront#1245] [Q&A]
I decided to try cpp2 and I'm experiencing issues initializing arrays of variables with value types. I have a type declared as
```
Point: @value type = {
    x: float;
    y: float;
    z: float;
}
```
and im trying to make an array of this within a function
```
corners: std::array<Point> = (
        ( 1, -1, -5),
        ( 1, -1, -3),
        ( 1,  1, -5),
        ( 1,  1, -3),
        (-1, -1, -5),
        (-1, -1, -3),
        (-1,  1, -5),
        (-1,  1, -3)
    );
```
I also tried prefixing each element with `Point` but i get an error saying "in operator=, expected 'x = ...' initialization statement (because type scope object 'x' does not have a default initializer)"

**Comment by Git-i** [comment]
I ended solving it like so:
```
Point: type = {
    public x: float;
    public y: float;
    public z: float;
    public operator=: (out this, xv: float, yv: float, zv: float) = {
        this.x = xv;
        this.y = yv;
        this.z = zv;
    }
}

main: () = {
    corners: std::array = (
        Point( 1, -1, -5),
        Point( 1, -1, -3),
        Point( 1,  1, -5),
        Point( 1,  1, -3),
        Point(-1, -1, -5),
        Point(-1, -1, -3),
        Point(-1,  1, -5),
        Point(-1,  1, -3)
    );
}
```

**Reply by MaxSagebaum** [reply]
Have you tried `@struct`? It generates the default constructor for all elements. Maybe this constructor generation should also be added to `@value`.

**Reply by hsutter** [reply]
Thanks, this is actually a bug report!

Yes, you can work around the problem by writing the code as you mentioned in the followup above.

But as @MaxSagebaum mentioned, `@struct` does pretty much everything you want (and is supposed to):

```cpp
Point: @struct type = {
    x: float;
    y: float;
    z: float;
}
```

... except there was a detail of how I was generating the memberwise constructor that disallowed implicit narrowing from integer literals or `double` floating point literals, so you would have to write things like `Point( 1.0f, -1.0f, -5.0f)` instead of `Point( 1, -1, -5)`. But with the change I'm about to push, this will work correctly:

```cpp
main: () = {
    corners: std::array = (
        Point( 1, -1, -5),  // works with next commit
        Point( 1, -1, -3),  // (previously required .0f on every literal)
        Point( 1,  1, -5),
        Point( 1,  1, -3),
        Point(-1, -1, -5),
        Point(-1, -1, -3),
        Point(-1,  1, -5),
        Point(-1,  1, -3)
    );
}
```
